mounts.go

```diff
@@ -529,49 +534,57 @@ func (f *FileMount) Unmount(task *TaskRun) error {
+			panic(fmt.Sprintf("Internal worker bug! Cannot calculate SHA256 of file %v that I have in my cache: %v", file, err))
+		}
+		if requiredSHA256 == "" {
+			task.Warnf("[mounts] Found existing download for %v with SHA256 %v, but since no required SHA256 is specified in task mounts, it is not possible to verify the integrity of this file", cacheKey, sha256)
 			return
 		}
-		fileCaches[cacheKey] = &Cache{
-			Location: file,
-			Hits:     1,
-			Created:  time.Now(),
-			Owner:    fileCaches,
-			Key:      cacheKey,
-			SHA256:   sha256,
+		if requiredSHA256 == sha256 {
+			task.Infof("[mounts] Found existing download for %v with (correct) SHA256 %v", cacheKey, sha256)
+			return
 		}
-		err = validateSHA256(fsContent.String(), sha256, fsContent.RequiredSHA256(), task)
-		return
+		task.Infof("Found existing download of %v with SHA256 %v but task definition explicitly requires %v so deleting it", cacheKey, sha256, requiredSHA256)
+		delete(fileCaches, cacheKey)
 	}
-	file = fileCaches[cacheKey].Location
-	// Sanity check - if file is in file map, but not on file system,
-	// something is seriously wrong, so should be a worker exception
-	// (panic), not a task failure
-	_, err = os.Stat(file)
+	file, sha256, err = fsContent.Download(task)
 	if err != nil {
-		panic(fmt.Errorf("File in cache, but not on filesystem: %v", *fileCaches[cacheKey]))
+		log.Printf("Could not download %v due to %v", fsContent.UniqueKey(), err)
+		return
 	}
-	fileCaches[cacheKey].Hits += 1
-
-	// validate SHA256 in case of tampering
-	sha256, err = fileutil.CalculateSHA256(file)
-	if err != nil {
-		panic(fmt.Sprintf("Internal worker bug! Cannot calculate SHA256 of file %v that I have in my cache: %v", file, err))
+	fileCaches[cacheKey] = &Cache{
+		Location: file,
+		Hits:     1,
+		Created:  time.Now(),
+		Owner:    fileCaches,
+		Key:      cacheKey,
+		SHA256:   sha256,
 	}
-	if requiredSHA256 := fsContent.RequiredSHA256(); requiredSHA256 != "" {
-		if requiredSHA256 != sha256 {
-			err = fmt.Errorf("[mounts] Found existing download of %v with SHA256 %v but task definition explicitly requires %v", cacheKey, sha256, requiredSHA256)
-		} else {
-			task.Infof("[mounts] Found existing download for %v with (correct) SHA256 %v", cacheKey, sha256)
-		}
-	} else {
-		task.Warnf("[mounts] Found existing download for %v with SHA256 %v, but since no required SHA256 is specified in task mounts, it is not possible to verify the integrity of this file", cacheKey, sha256)
+	if requiredSHA256 == "" {
+		task.Warnf("[mounts] Content from %v has SHA256 %v but task payload does not declare a required value, so content authenticity cannot be verified", fsContent, sha256)
+		return
+	}
+	if requiredSHA256 != sha256 {
+		err = fmt.Errorf("SHA256 of content just downloaded from %v was %v but task definition explicitly requires %v so deleting it and giving up (only one attempt is made if there are no connection failures)", fsContent, sha256, requiredSHA256)
```

Comment:
printing such a detailed error message looks great!  If it weren't a security risk, I'd love it if we could print out the first 50bytes in base64 to be able to do things like see that an S3 error message was downloaded instead of a binary file.  I don't think we should do this in public logs, though.